### PR TITLE
Explicit `LocaleFallbacker` argument for `DatagenDriver`

### PR DESCRIPTION
**Pull request by robertbastian**
Datagen implicitly requires a `LocaleFallbacker` for most configurations, so I'd rather make it explicit so that providers that don't expose fallback themselves are exportable.

I also removed the no-fallback mode, as that can (almost, up to `["und"]`) be modeled by the fallback path, and made the required settings arguments to `new`.

I also added a `supported_markers` method to the `ExportableDataProvider` trait; datagen will by default now support all available markers *for the provider*. This lets me remove the `all_markers` methods from the driver crate, which didn't make sense there. This allows removing the `all_markers`, `marker`, and `markers` methods from `icu_datagen`, so that it's registry-agnostic. `markers_for_bin` moves to `icu`, as that needs the `icu` crate anyway and only supports markers in `icu`.

### CODE REVIEW

File: provider/datagen/src/driver.rs [hunk 1]
```diff
@@ -354,115 +353,71 @@ impl FallbackOptions {
...
-    ///
-    /// If there are multiple [`LocaleFamily`]s for the same [`LanguageIdentifier`], the last entry
-    /// in the iterator takes precedence.
-    ///
-    /// [`langid!`]: crate::prelude::langid
-    pub fn with_locales_and_fallback(
-        mut self,
+    /// The fallbacker is used to resolve locale families, and to dedpulicate data if requested.
+    /// Make sure to use the same fallback data when loading from the provider at runtime.
+    pub fn new(
```
Comment (sffc) [inline provider/datagen/src/driver.rs:382]:
```suggestion
    /// The fallbacker is used to resolve locale families, and to dedpulicate data if requested.
    /// Make sure to use the same fallback data when loading from the provider at runtime.
    /// Commonly, you will export the fallback markers, in which case you should construct
    /// your fallbacker with the source provider (i.e. [`LocaleFallbacker::try_new_unstable`]).
    pub fn new(
```

File: provider/datagen/tests/test-options.rs [hunk 2]
```diff
@@ -753,61 +704,59 @@ fn explicit_preresolved_und() {
     // Should return the exact explicit locales set.
     assert_eq!(exported.keys().collect::<Vec<_>>(), locales);
 }
-
```
Comment (sffc) [inline provider/datagen/tests/test-options.rs:756]: Nit: Add back the newline

File: provider/fs/src/export/mod.rs [hunk 3]
```diff
@@ -25,11 +25,8 @@
 //! let mut exporter = FilesystemExporter::try_new(serializer, options)
 //!     .expect("Should successfully initialize data output directory");
 //!
-//! // Export something
-//! DatagenDriver::new()
-//!     .with_markers([HelloWorldV1Marker::INFO])
-//!     // HelloWorldProvider cannot provide fallback data, so we cannot deduplicate
-//!     .with_locales_and_fallback([LocaleFamily::FULL], FallbackOptions::no_deduplication())
+//! // Export something. Make sure to use the same fallback data at runtime.
```
Comment (sffc) [inline provider/fs/src/export/mod.rs:28]:
```suggestion
//! // Export something. Make sure to use the same fallback data at runtime!
```

File: provider/icu4x-datagen/src/main.rs [hunk 4]
```diff
@@ -295,16 +288,26 @@ fn main() -> eyre::Result<()> {
     let markers = if !cli.markers.is_empty() {
         match cli.markers.as_slice() {
             [x] if x == "none" => Default::default(),
-            [x] if x == "all" => icu_datagen::all_markers(),
+            [x] if x == "all" => {
+                #[cfg(feature = "experimental")]
+                log::warn!("The icu4x-datagen crate has been built with the `experimental` feature, so `--markers all` includes experimental markers");
+                #[cfg(not(feature = "experimental"))]
+                log::warn!("The icu4x-datagen crate has been built without the `experimental` feature, so `--markers all` does not include experimental markers");
```
Comment (sffc) [inline provider/icu4x-datagen/src/main.rs:295]: It's not really a warning, is it?

```suggestion
                #[cfg(feature = "experimental")]
                log::info!("The icu4x-datagen crate has been built with the `experimental` feature, so `--markers all` includes experimental markers");
                #[cfg(not(feature = "experimental"))]
                log::info!("The icu4x-datagen crate has been built without the `experimental` feature, so `--markers all` does not include experimental markers");
```

File: provider/icu4x-datagen/src/main.rs [hunk 5]
```diff
@@ -588,6 +580,74 @@ fn main() -> eyre::Result<()> {
...
+                .collect()
+            })
+        }
+
+        #[test]
+        fn test_lookup() {
+            assert_eq!(marker_lookup().get("AndListV1Marker"), Some(&Some(icu::list::provider::AndListV1Marker::INFO)));
+            assert_eq!(marker_lookup().get("list/and@1"), Some(&Some(icu::list::provider::AndListV1Marker::INFO)));
+            assert_eq!(marker_lookup().get("foo"), None);
+        }
```
Comment (sffc) [inline provider/icu4x-datagen/src/main.rs:630]: Nit: This test need not be inside the macro?

Comment (robertbastian) [reply provider/icu4x-datagen/src/main.rs:630]: I think it's nice to have it next to the function it tests

Comment (sffc) [reply provider/icu4x-datagen/src/main.rs:630]: I prefer if you move it 15 lines down and out of the macro

**Review by sffc — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Pull request merged**